datajoint/external.py

```diff
@@ -45,44 +45,82 @@ def __init__(self, connection, store=None, database=None):
         if not self.is_declared:
             self.declare()
 
+        self._s3 = None
+
     @property
     def definition(self):
         return """
         # external storage tracking
-        hash  : uuid
+        hash  : uuid    #  hash of contents (blob), of filename + contents (attach), or relative filepath (filepath)
         ---
-        size      :bigint unsigned   # size of object in bytes
+        size      :bigint unsigned     # size of object in bytes
+        filepath=null : varchar(1000)  # relative filepath used in the filepath datatype
+        contents_hash=null : uuid      # used for the filepath datatype 
         timestamp=CURRENT_TIMESTAMP  :timestamp   # automatic timestamp
         """
 
     @property
     def table_name(self):
         return '{external_table_root}_{store}'.format(external_table_root=EXTERNAL_TABLE_ROOT, store=self.store)
 
+    @property
+    def s3(self):
+        if self._s3 is None:
+            self._s3 = s3.Folder(**self.spec)
+        return self._s3
+
     def put(self, blob):
         """
-        put an object in external store
+        put a binary string in external store
         """
-        blob_hash = uuid_from_buffer(blob)
-        if self.spec['protocol'] == 'file':
-            folder = os.path.join(self.spec['location'], self.database, *subfold(blob_hash.hex, self.spec['subfolding']))
-            full_path = os.path.join(folder, blob_hash.hex)
-            if not os.path.isfile(full_path):
-                try:
-                    safe_write(full_path, blob)
-                except FileNotFoundError:
-                    os.makedirs(folder)
-                    safe_write(full_path, blob)
-        elif self.spec['protocol'] == 's3':
-            folder = '/'.join(subfold(blob_hash.hex, self.spec['subfolding']))
-            s3.Folder(database=self.database, **self.spec).put('/'.join((folder, blob_hash.hex)), blob)
-
+        uuid = uuid_from_buffer(blob)
+        if self.spec['protocol'] == 's3':
+            self.s3.put('/'.join((self.database, '/'.join(subfold(uuid.hex, self.spec['subfolding'])), uuid.hex)), blob)
+        else:
+            remote_file = os.path.join(os.path.join(
+                self.spec['location'], self.database, *subfold(uuid.hex, self.spec['subfolding'])), uuid.hex)
+            safe_write(remote_file, blob)
         # insert tracking info
         self.connection.query(
-            "INSERT INTO {tab} (hash, size) VALUES (%s, {size}) "
-            "ON DUPLICATE KEY UPDATE timestamp=CURRENT_TIMESTAMP".format(
-                tab=self.full_table_name, size=len(blob)), args=(blob_hash.bytes,))
-        return blob_hash
+            "INSERT INTO {tab} (hash, size) VALUES (%s, {size}) ON DUPLICATE KEY "
+            "UPDATE timestamp=CURRENT_TIMESTAMP".format(
```

Comment:
To simplify the management of granting user privileges, can we use `INSERT IGNORE` or `REPLACE INTO` here instead depending on what is the intended behavior? The former would require only `INSERT` privilege, the ladder would require `INSERT, DELETE`. The current requires `INSERT, UPDATE`.

Reply:
We use `ON DUPLICATE KEY UPDATE` as a way to handle duplicates without ignoring other types of errors as would be the case with `INSERT IGNORE`. This feature is also used in `Table.insert`. We have to give people `UPDATE` privilege to allow `skip_duplicates` functionality.